jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+from typing import Dict, cast
+
+import httpx
+from loguru import logger
+
+from .exceptions import AuthenticationError, TokenError
+from .token import Token, TokenType
+
+
+_LoginInformation = namedtuple(
+    "_LoginInformation",
+    ["verification_url", "wait_interval", "device_code"],
+)
+
+
+@dataclass
+class JobbergateAuth:
+    """
+    High-level class used to handle authentication in Jobbergate.
+
+    After an instance of this class is created, it can be used to authenticate requests
+    from both `requests`_  and `httpx`_ packages by passing it to the ``auth``
+    parameter on the request (see examples bellow).
+
+    It works just out of the box. Behind the scenes, this procedure calls the
+    :meth:`JobbergateAuth.acquire_tokens` method to load the available tokens from the cache directory, it
+    will try to refresh them if they are expired, or will provide an URL to the user to login on the system.
+
+    Notice all steps above are also available individually as public methods, allowing a fine control for
+    advanced users.
+
+    .. _requests: https://requests.readthedocs.io/en/latest/
+    .. _httpx: https://www.python-httpx.org/
+
+    Arguments:
+        cache_directory (Path): Directory to be used for the caching tokens.
+        login_domain (str): Domain used for the login.
+        login_audience (str): Audience of the login.
+        login_client_id (str): Client ID used for login.
+        tokens (Dict[TokenType, Token]): Dictionary holding the tokens needed for authentication.
+
+    Note:
+        Consult the values above with your system administrator.
+
+    Note:
+        This class can interoperate with the tokens generated by the ``jobbergate-cli`` package, as long
+        as they are stored in the same cache directory.
+
+    Examples:
+
+        The following example shows how to use the :meth:`JobbergateAuth` class to authenticate a request:
+
+        >>> from pathlib import Path
+        >>> import requests
+        >>> from jobbergate_core import JobbergateAuth
+        >>> jobbergate_auth = JobbergateAuth(
+        ...     cache_directory=Path("."),
+        ...     login_domain="http://keycloak.local:8080/realms/jobbergate-local",
+        ...     login_audience="https://local.omnivector.solutions",
+        ...     login_client_id="cli",
+        ... )
+        >>> jobbergate_base_url = "http://localhost:8000/jobbergate"
+        >>> jobbergate_auth.acquire_tokens()
+        Login Here: http://keycloak.local:8080/realms/jobbergate-local/device?user_code=LMVJ-XOLG
+        >>> response = requests.get(
+        ...     f"{jobbergate_base_url}/applications",
+        ...     auth=jobbergate_auth # this is the important part
+        )
+        >>> response.raise_for_status()
+        >>> print(f"response = {response.json()}")
+    """
+
+    cache_directory: Path
+    login_domain: str
+    login_audience: str
+    login_client_id: str
+    tokens: Dict[TokenType, Token] = field(default_factory=dict)
```

Comment:
Since there are only two token types, why not have each be its own attribute?
```python
access_token: Token
refresh_token: Token
```

I don't think using a dict here is helpful.

Reply:
I prefer a TypedDict here

Reply:
In terms of the number of code lines, they are not very different, I mean:

```python
        for token in self.tokens.values():
            token.save_to_cache()
```

takes the same two as:

```
        access_token.save_to_cache()
        refresh_token.save_to_cache()
```

One potential problem is that they can be nullable in the present implementation. For instance, just the access token is necessary to authenticate a request, and just the refresh token is necessary to refresh both of them. In addition, the Handler starts without any of them, `load_from_cache` of some other method is required to fill them in.

I think this is easier to handle in the dictionary.

Notice there is no need for end-users to access the tokens directly, this choice should not impact them. The Handler takes care of everything.

Reply:
I see no problem in removing it from initialization arguments and making it private `_tokens`.

Reply:
> In terms of the number of code lines, they are not very different, I mean:
> 
> ```python
>         for token in self.tokens.values():
>             token.save_to_cache()
> ```
> 
> takes the same two as:
> 
> ```
>         access_token.save_to_cache()
>         refresh_token.save_to_cache()
> ```


I understand what you are saying. However, you have the added complexity of a loop and a data structure (dict) when you only have two items and you know there won't be any additional items later. I think the overhead of using a data structure here for two elements is unnecessary and I would even call it a code smell.

Reply:
Ok, working on it